pkg/job/scraper_test.go

```diff
@@ -0,0 +1,454 @@
+			jobsCfg: model.JobsConfig{
+				DiscoveryJobs: []model.DiscoveryJob{
+					{
+						Regions: []string{"us-east-1"},
+						Type:    "aws-namespace",
+						Roles: []model.Role{
+							{RoleArn: "aws-arn-1", ExternalID: "external-id-1"},
+						},
+					},
+				},
+				CustomNamespaceJobs: []model.CustomNamespaceJob{
+					{
+						Regions:   []string{"us-east-2"},
+						Name:      "my-custom-job",
+						Namespace: "custom-namespace",
+						Roles: []model.Role{
+							{RoleArn: "aws-arn-2", ExternalID: "external-id-2"},
+						},
+					},
+				},
+			},
+			getAccountFunc: func() (string, error) {
+				return "", errors.New("failed to get account")
+			},
+			expectedErrs: []job.Error{
+				{JobContext: job.JobContext{AccountID: "", Namespace: "aws-namespace", Region: "us-east-1", RoleARN: "aws-arn-1"}},
+				{JobContext: job.JobContext{AccountID: "", Namespace: "custom-namespace", Region: "us-east-2", RoleARN: "aws-arn-2"}},
+			},
+		},
+		{
+			name: "returns errors from resource discovery without failing scrape",
+			jobsCfg: model.JobsConfig{
+				DiscoveryJobs: []model.DiscoveryJob{
+					{
+						Regions: []string{"us-east-1"},
+						Type:    "aws-namespace",
+						Roles: []model.Role{
+							{RoleArn: "aws-arn-1", ExternalID: "external-id-1"},
+						},
+					},
+				},
+				CustomNamespaceJobs: []model.CustomNamespaceJob{
+					{
+						Regions:   []string{"us-east-2"},
+						Name:      "my-custom-job",
+						Namespace: "custom-namespace",
+						Roles: []model.Role{
+							{RoleArn: "aws-arn-2", ExternalID: "external-id-2"},
+						},
+					},
+				},
+			},
+			getAccountFunc: func() (string, error) {
+				return "aws-account-1", nil
+			},
+			metadataRunFunc: func(_ context.Context, _ string, _ model.DiscoveryJob) ([]*model.TaggedResource, error) {
+				return nil, errors.New("I failed you")
+			},
+			cloudwatchRunFunc: func(_ context.Context, _ cloudwatchrunner.Job) ([]*model.CloudwatchData, error) {
+				return []*model.CloudwatchData{
+					{
+						MetricName:          "metric-2",
+						ResourceName:        "resource-2",
+						Namespace:           "custom-namespace",
+						Dimensions:          []model.Dimension{{Name: "dimension2", Value: "value2"}},
+						GetMetricDataResult: &model.GetMetricDataResult{Statistic: "Minimum", Datapoint: aws.Float64(2.0), Timestamp: time.Time{}},
+					},
+				}, nil
+			},
+			expectedMetrics: []model.CloudwatchMetricResult{
+				{
+					Context: &model.ScrapeContext{Region: "us-east-2", AccountID: "aws-account-1"},
+					Data: []*model.CloudwatchData{
+						{
+							MetricName:          "metric-2",
+							ResourceName:        "resource-2",
+							Namespace:           "custom-namespace",
+							Dimensions:          []model.Dimension{{Name: "dimension2", Value: "value2"}},
+							GetMetricDataResult: &model.GetMetricDataResult{Statistic: "Minimum", Datapoint: aws.Float64(2.0), Timestamp: time.Time{}},
+						},
+					},
+				},
+			},
+			expectedErrs: []job.Error{
```

Comment:
Could we match on the error message as well in these error cases tests?

Reply:
Ok I just read the comment later 

> We don't need the check the exact error or message

Why is that the case?

Reply:
I tend to avoid checking specific error messages unless it's pivital that an error show up with certain text it tends to make tests super brittle. If there's errors you think are really valuable I can type them and would feel better about checking types vs just text.

Reply:
Yeah checking types is a good option. I think that while at it, we could use this opportunity to start being more meticolous around errors that bubble up.

Reply:
I added this, https://github.com/nerdswords/yet-another-cloudwatch-exporter/pull/1432/files#diff-bf5be17ad3e7eb74bb4839c9cd988e47b9f3bc04a4debae6b12d3e0e24b3dad5R46-R52, which would be a way for us to give some of the more important use cases a type to match against. WDYT?